python/src/iceberg/types.py

```diff
@@ -15,61 +15,84 @@
 # specific language governing permissions and limitations
 # under the License.
 """Data types used in describing Iceberg schemas
-
 This module implements the data types described in the Iceberg specification for Iceberg schemas. To
 describe an Iceberg table schema, these classes can be used in the construction of a StructType instance.
-
 Example:
     >>> StructType(
         [
             NestedField(True, 1, "required_field", StringType()),
             NestedField(False, 2, "optional_field", IntegerType()),
         ]
     )
-
 Notes:
   - https://iceberg.apache.org/#spec/#primitive-types
 """
 
-from typing import Optional
+from typing import Any, Dict, List, Optional, Tuple
 
 
-class Type:
-    def __init__(self, type_string: str, repr_string: str, is_primitive=False):
-        self._type_string = type_string
-        self._repr_string = repr_string
-        self._is_primitive = is_primitive
+class IcebergType:
+
+    _implemented: Dict[Tuple[str, Tuple[Any]], "IcebergType"] = {}
+
+    def __new__(cls, *args, **kwargs):
```

Comment:
I can understand wanting to have one implementation for all classes, but I think that it is likely that the code would be more maintainable if it were split out because you wouldn't need to handle all possible values passed to `__init__`.

For example, adding this works for all of the non-parameterized primitive types:

```python
# not modified other than this addition
class Type:
    _instance = None
    def __new__(cls, *args, **kwargs):
        if not getattr(cls, "_instance"):
            cls._instance = super().__new__(cls, *args, **kwargs)
        return cls._instance
```
```python
In [4]: i = test_types.IntegerType()

In [5]: i == test_types.IntegerType()
Out[5]: True

In [6]: i == test_types.FloatType()
Out[6]: False

In [7]: test_types.FloatType()
Out[7]: FloatType()

In [8]: f = test_types.FloatType()

In [9]: f == test_types.FloatType()
Out[9]: True

In [10]: i == f
Out[10]: False
```

I think it would be much more straightforward to override `__new__` separately. Then because you control each `__init__` method, you know that you don't have to handle default args or kwargs or things that make this so complex.

Reply:
I think there's still a case to be made for a single implementation maybe just one that does not introspect and is more explicit? Say a decorator? 

I think something like 
```python

@singleton('length')
class FixedType(IcebergType):
# __new__ given via decorator
```
would be doable

`singleton` wouldn't have to do much but set a `new` that looks at what the decorator was fed. it would still have generic logic though just not introspection with the hidden attributes as it is currently. Do you think that's still too much?

Reply:
Also, I added some clarifying comments to the `__new__` in the latest commit. I think you already completely understand what's going on there though

Reply:
I agree with @rdblue about it. It would be much simpler and easy to understand and maintain if we do that separately. I gave it a try in https://github.com/apache/iceberg/pull/3965 and seems quite straightforward.

For the decorator example, while the object created using FixedType will be singleton, but IIUC, the FixedType itself is a function not a class any more, right? So we cannot call class methods from it, etc. It is better we don't use decorator.

Reply:
> For the decorator example, while the object created using FixedType will be singleton, but IIUC, the FixedType itself is a function not a class any more, right? So we cannot call class methods from it, etc. It is better we don't use decorator.

Decorators return a type they decorate. In this case, the decorator would take the class and return the very same class (in contrast to most decorators of functions which usually act as a closure and return a wrapper function)

Reply:
I ended up implementing `__new__` separately for all types that have some parameters and once for those that do not as the decorator would still have needed some resolving of arguments as it was in the single `__new__`